torch/_inductor/pcache.py

```diff
@@ -0,0 +1,315 @@
+
+        Returns:
+            bool: True if the value was stored successfully, False if the key already exists.
+
+        Raises:
+            NotImplementedError: If not implemented by a subclass.
+        """
+        raise NotImplementedError
+
+
+class InMemoryCache(Cache[str, bytes]):
+    """
+    In-memory cache implementation.
+
+    Stores cache data in a dictionary for fast lookups and insertions.
+    Not thread-safe.
+    """
+
+    def __init__(self: Self) -> None:
+        """
+        Initialize the in-memory cache.
+        """
+        self._cache: dict[str, bytes] = {}
+
+    def get(self: Self, key: str) -> bytes | None:
+        """
+        Retrieve the value associated with the given key from the cache.
+
+        Args:
+            key (str): The key used to query the cache.
+
+        Returns:
+            bytes | None: The value associated with the key, or None if not found.
+        """
+        return self._cache.get(key, None)
+
+    def insert(self: Self, key: str, value: bytes) -> bool:
+        """
+        Store the given value in the cache with the associated key if the key does
+        not already exist in the cache, otherwise do nothing.
+
+        Args:
+            key (str): The key to associate with the value.
+            value (bytes): The value to be stored in the cache.
+
+        Returns:
+            bool: True if the value was stored successfully, False if the key already exists.
+        """
+        if key in self._cache:
+            return False
+        self._cache[key] = value
+        return True
+
+    @classmethod
+    def from_env_var(cls, env_var: str) -> Self:
+        """
+        Create a new in-memory cache instance from an environment variable.
+
+        The environment variable should contain key-value pairs separated by ';',
+        with each pair formatted as 'key,value'. The value should be a string
+        representation of bytes (e.g., b'...').
+
+        Args:
+            env_var (str): The environment variable containing cache data.
+
+        Returns:
+            InMemoryCache: A new in-memory cache instance populated with data from the environment variable.
+
+        Raises:
+            ValueError: If a key is associated with two distinct values, or if the environment variable
+                is malformed (e.g., missing comma, value not a bytes string).
+        """
+        cache: Self = cls()
+        env_val: str | None = getenv(env_var, None)
+
+        if env_val is not None:
+            for kv_pair in env_val.split(";"):
+                if kv_pair == "":
+                    # can happen if env_val is an empty string, or ends with ;
+                    continue
+                try:
+                    key, str_value = kv_pair.split(",")
+                except ValueError as err:
+                    raise ValueError(
+                        f"Malformed kv_pair {kv_pair} in env_var {env_var}!"
```

Comment:
nit
I would report the string as repr and also indicate a little more "why":
```suggestion
                        f"Malformed kv_pair {kv_pair!r} in env_var {env_var} (missing comma)!"
```

Reply:
Gotcha, fixed this and the one below